### ISSUE TITLE
SFT Llama 2 chat model ignoring EOS and BOS tokens

### ISSUE THREAD

**Issue by mallorbc** [huggingface/trl#696]
Hello.

Using SFTTrainer, and Qlora, I have been finetuning a variety of LLama 2 Chat models.  I have my dataset structured like the following based on what I have read to be the correct format:
```
[INST]<<SYS>>
You are a friendly chatbot that talks like a pirate
<</SYS>>

Hello[/INST]Arg how do you be?</s><s>[INST]Good, I am after your treasure![/INST]Arg!</s>
```

When using a vanilla model, this format works great.  The model follows the system prompt and conversation history and stops generating at the EOS token.

However, after finetuning on a dataset following the same pattern, the model no longer generates EOS tokens and instead keeps generating the assistant output in a format like this, often repeating:

```
[INST]<<SYS>>
You are a friendly chatbot that talks like a pirate
<</SYS>>

Hello[/INST]Arg how do you be?[/INST]Arg how do you be?[/INST]Arg how do you be?[/INST]Arg how do you be?
```

I have gone inside the code and printed out the input_ids and the decoded input_ids from the dataset after the train dataset is prepared here, and I see no change in the text, such as missing EOS and BOS tokens.  The only difference I see is an added BOS token to the beginning of the sequence.
https://github.com/huggingface/trl/blob/d3bbee3ab84d4d5dcb413003bf93541df5b4e54c/trl/trainer/sft_trainer.py#L182-L192

Thus, I am assuming that somewhere deeper in the code, the EOS and BOS tokens that normally are not in the middle of the sequence are being removed during training.  I am hoping that there is a way I can avoid this and get the proper behavior.

**Comment by mallorbc** [comment]
Still working on this, but I think it's due to the fact that there is no pad token set, and by default, its then using the EOS token, which normally is fine, but it's not really an EOS in this case.

https://www.reddit.com/r/LocalLLaMA/comments/15hz7gl/my_finetuning_based_on_llama27bchathf_model/

Currently training a model using the dirty fix of using a rare token for padding, but that is not a great solution.  Could also add a new token, but I have read there are issues doing so with Lora. Additionally changing the embedding dimensions for this is not great in my opinion.

I opened an issue here https://github.com/huggingface/transformers/issues/25773

In Meta's implementation, they use -1 for the pad token id, which is fine since padding is just for the attention mask.  Allowing -1 to be set for pad token seems to be the most elegant solution.

I will update any more findings here.

**Comment by mallorbc** [comment]
Using the dirty fix is definitely working.   Adding a new pad token probably works as well, but I have not tried it.  May try it in the near future.

**Comment by lvwerra** [comment]
Since the fix is working and this is a `transformers` issue I am closing the the issue for now. Feel free to re-open should there be an update. Thanks a lot for reporting!

**Comment by RonanKMcGovern** [comment]
Thanks @mallorbc , really interesting.

A few thoughts/questions:

1. What are you using as the rare token?

2. I believe that there is an attention mask AND a loss mask of 0s set for pad tokens, so if you set the pad token to the eos token then the eos token will get zerod out for attention, and potentially for loss. [Can someone confirm whether there is a loss mask used in SFTTrainer for pad tokens?]

3. My typical approach is to set the pad token to < pad >, see [here](https://github.com/TrelisResearch/llama-2-setup) and increase the size of the vocab for the tokenizer and model. I don't find any specific issues with LoRA - although I'd be interested in reading more on that if you have a link.

4. My specific issue is using SFTTrainer [issue here](https://github.com/huggingface/trl/issues/837) where I have set a new pad token of < pad >, but the fine-tuned model is not emitting EOS tokens as I would expect. Appreciate any insights you may have there.

**Comment by RonanKMcGovern** [comment]
Closing this off.

I don't fully understand the attention mask in the trainer, but given I'm not using EOS as the pad token, I don't believe there is an issue.

Furthermore, the issue I was having with the EOS token not being emitted seems to have been specific with the model I was using (TinyLlama), which in any case wasn't consistently pre-trained with an EOS. I didn't have the issue with other models like Llama 7B base.
